persistence/postgres-persistence/src/main/java/com/netflix/conductor/postgres/dao/PostgresLockDAO.java

```diff
@@ -0,0 +1,130 @@
+
+        // execute the query
+        queryWithTransaction(ADVISORY_LOCK, (q) -> {
+            q.addParameter(lockId);
+            q.executeQuery();
+
+            return q;
+        });
+    }
+
+    @Override
+    public boolean acquireLock(String lockId, long timeToTry, TimeUnit unit) {
+        final String TRY_ADVISORY_LOCK =
+                "DECLARE\n" +
+                        "  lock_acquired boolean;\n" +
+                        "  start_time timestamp;\n" +
+                        "BEGIN\n" +
+                        "  start_time := now();\n" +
+                        "\n" +
+                        "  lock_acquired := pg_try_advisory_lock(?);\n" +
+                        "\n" +
+                        "WHILE NOT lock_acquired AND start_time + INTERVAL '? ?' < now() LOOP\n" +
+                        "  PERFORM pg_sleep(0.5);\n" +
+                        "  lock_acquired := pg_try_advisory_lock(?);\n" +
+                        "END LOOP;\n" +
+                        "\n" +
+                        "SELECT lock_acquired;\n" +
+                        "\n" +
+                        "END;";
+
+        Query query = queryWithTransaction(TRY_ADVISORY_LOCK, (q) -> {
+            q.addParameter(lockId);
+            q.addParameter(timeToTry);
+            q.addParameter(unit.toString());
+            q.addParameter(lockId);
+
+            return q;
+        });
+
+        return query.executeAndFetchFirst(Boolean.class);
+    }
+
+    @Override
+    public boolean acquireLock(String lockId, long timeToTry, long leaseTime, TimeUnit unit) {
+        final String TRY_ADVISORY_LOCK =
+                "DECLARE\n" +
+                        "  lock_acquired boolean;\n" +
+                        "  start_time timestamp;\n" +
+                        "BEGIN\n" +
+                        "  start_time := now();\n" +
+                        "\n" +
+                        "  lock_acquired := pg_try_advisory_lock(?);\n" +
+                        "\n" +
+                        "WHILE NOT lock_acquired AND start_time + INTERVAL '? ?' < now() LOOP\n" +
+                        "  PERFORM pg_sleep(0.5);\n" +
+                        "  lock_acquired := pg_try_advisory_lock(?);\n" +
+                        "END LOOP;\n" +
+                        "\n" +
+                        "SELECT lock_acquired;\n" +
+                        "\n" +
+                        "END;";
+
+        Query query = queryWithTransaction(TRY_ADVISORY_LOCK, (q) -> {
+            q.addParameter(lockId);
+            q.addParameter(timeToTry);
+            q.addParameter(unit.toString());
+            q.addParameter(lockId);
+            q.executeQuery();
+
+            return q;
+        });
+
+        SCHEDULEDFUTURES.put(
+                lockId, SCHEDULER.schedule(() -> deleteLock(lockId), leaseTime, unit)
+        );
+
+        return query.executeAndFetchFirst(Boolean.class);
+    }
+
+    @Override
+    public void releaseLock(String lockId) {
```

Comment:
This service needs to be "reentrant" meaning conductor can call multiple acquireLock() and multiple releaseLock() for the same lock.

So what can happen is:
`
acquireLock("x", 10, 20, SECONDS);
acquireLock("x", 10, 20, SECONDS);
acquireLock("x", 10, 20, SECONDS);

releaseLock("x");
`

According to psql docs:
`Multiple session-level lock requests stack, so that if the same resource identifier is locked three times there must then be three unlock requests to release the resource in advance of session end.`

we should call release as many times as we called acquire. But conductor might not call release for each acqiure. In case something fails, it relies on the "leaseTime" to release the lock. But we remove any scheduled releases when `releaseLock` is called. In fact we don't even support multiple auto release futures properly as the SCHEDULEDFUTURES map only allows 1 future to be mapped to 1 future.

Reply:
WE need to properly test this thoroughly.